### Instructions:

- You can attempt any number of questions and in any order.  
  See the assignment page for a description of the hurdle requirement for this assessment.
- You may submit your practical for autograding as many times as you like to check on progress, however you will save time by checking and testing your own code before submitting.
- Develop and check your answers in the spaces provided.
- **Replace** the code `raise NotImplementedError()` with your solution to the question.
- Do **NOT** remove any variables other provided markings already provided in the answer spaces.
- Do **NOT** make any changes to this notebook outside of the spaces indicated.  
  (If you do this, the submission system might not accept your work)

### Submitting:

1. Before you turn this problem in, make sure everything runs as expected by resetting this notebook.    
   (You can do this from the menubar above by selecting `Kernel`&#8594;`Restart Kernel and Run All Cells...`)
1. Don't forget to save your notebook after this step.
1. Submit your .ipynb file to Gradescope via file upload or GitHub repository.
1. You can submit as many times as needed.



---

# <mark style="background: #801010; color: #ffffff;" >B2</mark> Topic 5: NumPy Arrays

#### Question 01 &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;(2 Points)

Create a python program that has a 1-D array with 5 elements. It should contain the numbers 0 to 4 in sequence. Assign this array to `output1`.

In [2]:
import numpy as np

# Write your solution here

# YOUR CODE HERE
output1 = np.arange(5)
print(output1)

[0 1 2 3 4]


In [ ]:
# Testing Cell (Do NOT modify this cell)

#### Question 02 &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;(2 Points)

Create a python program that contains a 2-D array with 6 elements (2 rows and 3 columns) that represents the following array. Assign this array to `output2`.

$$\begin{bmatrix} 0 & 1 & 2 \\ 3 & 4 & 5 \end{bmatrix}$$

In [5]:
import numpy as np

# Write your solution here

# YOUR CODE HERE
output2 = np.arange(6).reshape(2,3)
print(output2)
# output2 = np.array([[0,1,2], [3,4,5]])
# print(output2)

[[0 1 2]
 [3 4 5]]


In [ ]:
# Testing Cell (Do NOT modify this cell)

#### Question 03 &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;(2 Points)
Create a python program that has a 2-D array with 6 elements (3 rows and 2 columns) that represents the following array. Assign this array to `output3`.

$$\begin{bmatrix} 0 & 1 \\ 2 & 3 \\ 4 & 5 \end{bmatrix}$$

In [6]:
import numpy as np

# Write your solution here

# YOUR CODE HERE
output3 = np.arange(6).reshape(3,2)
print(output3)

[[0 1]
 [2 3]
 [4 5]]


In [ ]:
# Testing Cell (Do NOT modify this cell)

#### Question 04 &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;(2 Points)

Create a python program that has a 3-D array with 8 elements (2 rows and 2 columns, with a depth of 2) using the `zeros` function. Assign this array to `output4`.

In [8]:
import numpy as np

# Write your solution here

# YOUR CODE HERE
output4 = np.zeros((2,2,2))
print(output4)

[[[0. 0.]
  [0. 0.]]

 [[0. 0.]
  [0. 0.]]]


In [ ]:
# Testing Cell (Do NOT modify this cell)

#### Question 05 &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;(2 Points)

You are given a 2-D array of unknown size named `array5`. Create a Python function that finds the size of that array and returns the number of rows.

In [11]:
import numpy as np

def question05(array5):

    # Write your solution here

    # YOUR CODE HERE
    size = np.size(array5)
    row, column = np.shape(array5)
    
    return row
array = np.array([[0,1,2], [3,4,5]])
print(question05(array))

(6, 2)


In [ ]:
# Testing Cell (Do NOT modify this cell)

#### Question 06 &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;(2 Points)

Create a python program that performs the following calculation using universal functions. Assign the result to `output6`.

$$\begin{bmatrix} 0 & 1 \\ 2 & 3 \end{bmatrix} + \begin{bmatrix} 0 & 1 \\ 2 & 3 \end{bmatrix}$$

In [15]:
import numpy as np

# Write your solution here

# YOUR CODE HERE
array = np.arange(4).reshape(2,2)
output6 = array + array
print(output6)

[[0 2]
 [4 6]]


In [ ]:
# Testing Cell (Do NOT modify this cell)

#### Question 07 &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;(4 Points)
You are given two 2-D arrays of that are the same size (`array7_i` and `array7_ii`). Create a python program that multiplies the two arrays together using universal functions. Return a new array containing the result.

In [16]:
import numpy as np

def question07(a1, a2):

    # Write your solution here
    output7 = np.multiply(a1, a2)
    return output7
    # YOUR CODE HERE
array7_i = np.array([[323,123], [323,443]])
array7_ii = np.array([[423,13232], [321,3245]])
print(question07(array7_i, array7_ii))

[[ 136629 1627536]
 [ 103683 1437535]]


In [ ]:
# Testing Cell (Do NOT modify this cell)

#### Question 08 &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;(4 Points)

Create a python program that performs the following calculation using universal functions. Assign the result to `output8`.

$$\begin{bmatrix} 0 & 1 \\ 2 & 3 \end{bmatrix} ** 3$$

In [17]:
import numpy as np

# Write your solution here

# YOUR CODE HERE
array = np.arange(4).reshape(2,2)
output8 = np.power(array, 3)
print(output8)

[[ 0  1]
 [ 8 27]]


In [ ]:
# Testing Cell (Do NOT modify this cell)

#### Question 09 &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;(4 Points)

Create a python program that has a 1-D array with 100 elements. It should contain the numbers 0, 2, 4, 6, all the way to 198 in sequence. Assign this array to `output9`.

In [21]:
import numpy as np

# Write your solution here

# YOUR CODE HERE
output9 = np.arange(0,200,2)
print(output9)

[  0   2   4   6   8  10  12  14  16  18  20  22  24  26  28  30  32  34
  36  38  40  42  44  46  48  50  52  54  56  58  60  62  64  66  68  70
  72  74  76  78  80  82  84  86  88  90  92  94  96  98 100 102 104 106
 108 110 112 114 116 118 120 122 124 126 128 130 132 134 136 138 140 142
 144 146 148 150 152 154 156 158 160 162 164 166 168 170 172 174 176 178
 180 182 184 186 188 190 192 194 196 198]


In [ ]:
# Testing Cell (Do NOT modify this cell)

#### Question 10 &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;(4 Points)

Create a python program that has a 1-D array with 16 elements. It should contain the numbers 10, 20, 30, 40, all the way to 160 in sequence. Afterwards, change its shape using `reshape` to a 4-D array. Assign this reshaped array to `output10`.

***
<details>
  <summary><span style="color:blue">Click to see Hints</span></summary>
   In one of the provided examples, you saw how a 2-D 3-by-3 array. Use that as a basis for reshaping into a 4-D 2-by-2-by-2-by-2 array. 
</details>

In [25]:
import numpy as np

# Write your solution here

# YOUR CODE HERE
output10 = np.arange(10, 161, 10).reshape(2,2,2,2)
print(output10)

[[[[ 10  20]
   [ 30  40]]

  [[ 50  60]
   [ 70  80]]]


 [[[ 90 100]
   [110 120]]

  [[130 140]
   [150 160]]]]


In [ ]:
# Testing Cell (Do NOT modify this cell)

#### Question 11 &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;(6 Points)

One of the most common models in statistics (and by extention data science) is the normal distribution. In order to graph a normal distribution first the mean and standard deviation of the data must be found. NumPy has functions that allow us to analyse data and obtain these values.

You are given an array of unknown size and dimensions named `array11` that represents some data. Create a python program that finds the both the mean and standard deviation (rounded to 4 decimal points) of the data. Return the values in a dictionary of the following structure:
```python
question11 = {
  "mean": some_value,
  "standard_deviation": some_value
}
```

In [16]:
import numpy as np

def question11(array11):
    mean = round((np.mean(array11)), 4)
    deviation =round((np.std(array11)),4)
    output11 = {
        'mean': mean,
        'standard_deviation':deviation
    }
    return output11


array11 = np.random.randn(100)

print(question11(array11))



{'mean': -0.0579, 'standard_deviation': 0.9008}


In [ ]:
# Testing Cell (Do NOT modify this cell)

#### Question 12 &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;(6 Points)

You are given an 1-D array of unknown size named `array12`. Create a python program that finds both the first element and the minimum element in the array determined using a standard function. Return the values in a dictionary of the following structure:
```python
{
  "first": some_value,
  "smallest": some_value
}
```

In [23]:
import numpy as np

def question12(array12):

    first = array12[0]
    smallest = np.min(array12)
    output12 = {
        "first": first,
        "smallest": smallest
    }
    return output12
array12 = np.random.randint(1, 100, size = 5)
print(array12)
print(question12(array12))

[31 26 81 31 56]
{'first': 31, 'smallest': 26}


In [ ]:
# Testing Cell (Do NOT modify this cell)

#### Question 13 &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;(6 Points)

You are given an 2-D array of unknown size named `array13`. Create a python program that finds the first row of data and the maximum number within that row.  Return the values in a dictionary of the following structure:
```python
question13 = {
  "first_row": some_value,
  "max_row_value": some_value
}
```

In [56]:
import numpy as np

def question13(array13):

    first = array13[0]
    max = np.max(first)
    output13 = {
        "first_row": first,
        "max_row_value": max
    }
    return output13

arrays = np.random.randint(1,100, size=(5,5))
print(question13(arrays))    

{'first_row': array([59, 26, 12, 49, 61]), 'max_row_value': 61}


In [ ]:
# Testing Cell (Do NOT modify this cell)


#### Question 14 &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;(6 Points)

You are given an 1-D array of unknown size named `array14`. Create a python program that finds the first 6 entries within that row. Next, split that array into two arrays using the `split` function. Return the values in a dictionary of the following structure:
```python
question14 = {
  "array": first_6_entries,
  "part1": split_first_half,
  "part2": split_other_half
}
```

In [59]:
import numpy as np

def question14(array14):

    # Write your solution here

    # YOUR CODE HERE
    array = array14[:6]
    part1, part2 = np.split(array, 2)
    output14 = {
        "array": array,
        "part1": part1,
        "part2": part2
    }
    return output14
array14 = np.random.randint(1,100, size=10)
print(question14(array14))

{'array': array([29, 86, 99, 40, 82,  3]), 'part1': array([29, 86, 99]), 'part2': array([40, 82,  3])}


In [ ]:
# Testing Cell (Do NOT modify this cell)

#### Question 15 &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;(6 Points)

You are given an 1-D array of 40 elements named `array15`. Create a Python program that finds every second element starting from index 4 and stopping before index 20. Return the result as an array.

In [61]:
import numpy as np

def question15(array15):

    # Write your solution here

    # YOUR CODE HERE
    output15 = array15[4:20:2]
    return output15
array15 = np.random.randint(1,100, size=40)
print(array15)
print(question15(array15)) 

[45 86 19 20 61 22 36 32 23 51 79 23 26 32 53 29 99 19 77 88 40 63 98 90
 28 56 90 38 11 38  9 25 64 60 67  9 94 16 24 29]
[61 36 23 79 26 53 99 77]


In [ ]:
# Testing Cell (Do NOT modify this cell)

#### Question 16 &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;(6 Points)

You are given 2 sorted 1-D arrays of 20 elements named `array16_i` and `array16_ii`. Create a python program that concatenates these two arrays by appending `array16_ii` to the end of `array16_i`. Afterwards, sort the new array.
Return both the unsorted concatenated array, and the sorted array in a dictionary of the following structure:
```python
question16 = {
  "concat": concatenated_array,
  "sorted": sorted_array
}
```

In [67]:
import numpy as np

def question16(array16_i, array16_ii):

    # Write your solution here

    # YOUR CODE HERE
    concat = np.concatenate([array16_i, array16_ii])
    sorted_array = np.sort(concat)
    output16 = {
        "concat": concat,
        "sorted": sorted_array
    }
    return output16
array16_i = np.sort(np.random.randint(1,10, size = 20))
array16_ii = np.sort(np.random.randint(1,10, size = 20))
print(question16(array16_i, array16_ii))

{'concat': array([1, 2, 2, 4, 4, 4, 5, 5, 5, 5, 6, 6, 6, 7, 7, 8, 8, 9, 9, 9, 1, 2,
       2, 3, 3, 3, 3, 3, 3, 4, 6, 7, 7, 8, 8, 8, 9, 9, 9, 9]), 'sorted': array([1, 1, 2, 2, 2, 2, 3, 3, 3, 3, 3, 3, 4, 4, 4, 4, 5, 5, 5, 5, 6, 6,
       6, 6, 7, 7, 7, 7, 8, 8, 8, 8, 8, 9, 9, 9, 9, 9, 9, 9])}


In [ ]:
# Testing Cell (Do NOT modify this cell)

#### Question 17 &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;(8 Points)

Create a python program that:

1. initializes two arrays using the `arange` function followed by the `reshape` functions.
   1. The first array should have 2 rows and 5 columns and contain the numbers 0, 1, 2, up to and including 9.
   2. The second array should  have 2 rows and 3 columns and contain the numbers 1, 3, 5, up to and including 11.
2. concatenates these two arrays by appending along the column axis (creating an array with 2 rows and 8 columns), then transposes the result (creating an array with 8 rows and 2 columns).

Assign the end result to `output17`.

In [45]:
# Write your solution here

# YOUR CODE HERE
import numpy as np
first_array = np.arange(10).reshape(2,5)
second_array = np.arange(1,12,2).reshape(2,3)
output17 = np.concatenate([first_array, second_array], 1)
print(output17)
output17 = output17.T

print(output17)



[[ 0  1  2  3  4  1  3  5]
 [ 5  6  7  8  9  7  9 11]]
[[ 0  5]
 [ 1  6]
 [ 2  7]
 [ 3  8]
 [ 4  9]
 [ 1  7]
 [ 3  9]
 [ 5 11]]


In [ ]:
# Testing Cell (Do NOT modify this cell)

#### Question 18 &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;(8 Points)

Create a python program that:

1. initializes two arrays using the `arange` function followed by the `reshape` functions.
   1. The first array should have 4 rows and 2 columns and contain the numbers 0, 1, 2, up to 7.
   2. The second array should have 3 rows and 2 columns and contain the numbers 1, 3, 5, up to 11.
2. concatenates these two arrays by appending along the row axis.
3. afterwards, reshape the new array into a 1-D array and sort it in decending order.

Assign the end result to `output18`.

In [22]:
# Write your solution here

# YOUR CODE HERE
first_array = np.arange(8).reshape(4,2)
second_array = np.arange(1,12,2).reshape(3,2)
output18 = np.concatenate([first_array, second_array], 0)
output18 = output18.reshape(-1)
output18 =np.sort(output18)[::-1]
print(output18)

[11  9  7  7  6  5  5  4  3  3  2  1  1  0]


In [ ]:
# Testing Cell (Do NOT modify this cell)

#### Question 19 &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;(10 Points)

Create a python program that:

1. initializes two 2-D arrays as depicted below.
$$a1 = \begin{bmatrix} 70 & 51 & 89 \\ 42 & 15 & 8 \end{bmatrix}, a2 = \begin{bmatrix} 75 & 12 & 38 \\ 99 & 75 & 45 \end{bmatrix}$$
2. performs integer division, by dividing the elements of a1 by a2. Assign this result to `output19_i`.
3. sums the three arrays together. Assign this result to `output19_ii`, also assign a copy of the second row to `output19_iii`.


In [28]:
# Write your solution here

# YOUR CODE HERE
a1 = np.array([[70,51,89],[42,15,8]])
a2 = np.array([[75, 12, 38], [99, 75, 45]])
output19_i = a1 // a2
output19_ii = a1 + a2 + output19_i
print(output19_i)
print(output19_ii)
output19_iii = output19_ii[1].copy()
print(output19_iii)

[[0 4 2]
 [0 0 0]]
[[145  67 129]
 [141  90  53]]
[141  90  53]


In [ ]:
# Testing Cell (Do NOT modify this cell)

#### Question 20 &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;(10 Points)

Create a python program that:

1. initializes two 2-D arrays as depicted below.
$$a1 = \begin{bmatrix} 48 & 75 & 75 \\ 15 & 2 & 46 \end{bmatrix}, a2 = \begin{bmatrix} 78 & 15 & 87 \\ 0 & 61 & 9 \end{bmatrix}$$
2. multiplies the two arrays together, then slice the result to only assign the last column to `output20` (note: an n-row, 1-column array automatically converts into a 1-row, n-column array, leave your answer like this).


In [1]:
    # Write your solution here

    # YOUR CODE HERE

a1 = np.array([[48, 75, 75], [15, 2, 46]])
a2 = np.array([[78, 15, 87], [0, 61, 9]])

output20 = a1 * a2
print(output20)
last_column = output20[:, -1]
print('---')
output20 = last_column
print(output20)

NameError: name 'np' is not defined

In [ ]:
# Testing Cell (Do NOT modify this cell)